rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -158,237 +75,165 @@ def _create_and_append_fee_event(
+        # the rainbow router
+        if self.evm_inquirer.native_token in {out_event.asset, in_event.asset}:
+            for internal_tx in DBEvmTx(self.evm_inquirer.database).get_evm_internal_transactions(
+                parent_tx_hash=transaction.tx_hash,
+                blockchain=self.evm_inquirer.blockchain,
             ):
-                if bytes_to_address(tx_log.topics[2]) == self.router_address:
-                    sell_amount = int.from_bytes(tx_log.data)
-                elif bytes_to_address(tx_log.topics[1]) == self.router_address:
-                    swap_amount = int.from_bytes(tx_log.data)
+                if ((
+                    internal_tx.from_address == self.router_address and
+                    internal_tx.to_address != out_event.location_label
+                ) or (
+                    internal_tx.from_address != out_event.location_label and
+                    internal_tx.to_address == self.router_address
+                )):
+                    swapped_amount = from_wei(FVal(internal_tx.value))
+                    break
 
-                if swap_amount and sell_amount:
+        # When eth is swapped the fee is taken from the sent ETH. When ETH is received the
+        # fee is taken from the received amount. The fee is the amount diff with the amount
+        # that the proxy sends to the swap solver
+        fee_asset = fee_amount = None
+        if (
+            out_event.asset == self.evm_inquirer.native_token and
+            (possible_fee_amount := out_event.amount - swapped_amount) > 0
+        ):
+            fee_amount = possible_fee_amount
+            out_event.amount -= fee_amount
+            fee_asset = self.evm_inquirer.native_token
+        elif (
+            in_event.asset == self.evm_inquirer.native_token and
+            (possible_fee_amount := swapped_amount - in_event.amount) > 0
+        ):
+            fee_amount = possible_fee_amount
+            fee_asset = self.evm_inquirer.native_token
+        else:  # token transfer
+            # check in the logs the transfers made by the router to find if the fee was
+            # taken from the asset sent or the asset received
+            for log_event in all_logs:
+                if len(log_event.topics) != 3 or log_event.topics[0] == ERC20_OR_ERC721_TRANSFER:
+                    continue  # we look only for transfers
+
+                if (
+                    bytes_to_address(log_event.topics[1]) == self.router_address and
+                    bytes_to_address(log_event.topics[2]) != in_event.location_label and
+                    (possible_fee_amount := (out_event.amount - asset_normalized_value(
+                        amount=int.from_bytes(log_event.data),
+                        asset=(out_asset := out_event.asset.resolve_to_crypto_asset()),
+                    ))) > 0
+                ):
+                    # this is the router sending to the solver the token and it charged a fee
```

Comment:
save lines, plus rewording

```suggestion
                ):  # this is the router sending the token to the solver and charging a fee
```